index.js

```diff
@@ -786,7 +786,10 @@
 
   //  Array$prototype$extend :: Array a ~> (Array a -> b) -> Array b
   function Array$prototype$extend(f) {
-    return [f(this)];
+    var xs = this;
+    return xs.map(function(_, i) {
+      return f(xs.slice(i));
+    });
```

Comment:
@davidchambers this would be definitely better, but client might store the all `xs`es and we would mutate it.
in PS List is Extend and it's linear as it's going from right to left: https://github.com/purescript/purescript-lists/blob/v4.0.1/src/Data/List/Lazy/Types.purs#L161-L170

but we it would make no sense for array so this version is fine (btw there is no extend instance for array in PS)